rasa/nlu/extractors/crf_entity_extractor.py

```diff
@@ -130,216 +150,142 @@ def train(
-                logger.debug(
-                    "Inconsistent BILOU tagging found, B- tag, L- "
-                    "tag pair encloses multiple entity classes.i.e. "
-                    "[B-a, I-b, L-a] instead of [B-a, I-a, L-a].\n"
-                    "Assuming B- class is correct."
-                )
+        predictions = {}
+        for tag_name, entity_tagger in self.entity_taggers.items():
+            # use predicted entity tags as features for second level CRFs
+            include_tag_features = tag_name != ENTITY_ATTRIBUTE_TYPE
+            if include_tag_features:
+                self._add_tag_to_crf_token(crf_tokens, predictions)
 
-            if label.startswith("L-"):
-                # end of the entity
-                finished = True
-            elif label.startswith("I-"):
-                # middle part of the entity
-                ent_word_idx += 1
-            else:
-                # entity not closed by an L- tag
-                finished = True
-                ent_word_idx -= 1
-                logger.debug(
-                    "Inconsistent BILOU tagging found, B- tag not "
-                    "closed by L- tag, i.e [B-a, I-a, O] instead of "
-                    "[B-a, L-a, O].\nAssuming last tag is L-"
-                )
-        return ent_word_idx, confidence
+            features = self._crf_tokens_to_features(crf_tokens, include_tag_features)
+            predictions[tag_name] = entity_tagger.predict_marginals_single(features)
 
-    def _handle_bilou_label(
-        self, word_idx: int, entities: List[Any]
-    ) -> Tuple[Any, Any, Any]:
-        label, confidence = self.most_likely_entity(word_idx, entities)
-        entity_label = bilou_utils.entity_name_from_tag(label)
+        # convert predictions into a list of tags and a list of confidences
+        tags, confidences = self._tag_confidences(tokens, predictions)
 
-        if bilou_utils.bilou_prefix_from_tag(label) == "U":
-            return word_idx, confidence, entity_label
+        return self.create_entities(message.text, tokens, tags, confidences)
 
-        elif bilou_utils.bilou_prefix_from_tag(label) == "B":
-            # start of multi word-entity need to represent whole extent
-            ent_word_idx, confidence = self._find_bilou_end(word_idx, entities)
-            return ent_word_idx, confidence, entity_label
+    def _add_tag_to_crf_token(
+        self,
+        crf_tokens: List[CRFToken],
+        predictions: Dict[Text, List[Dict[Text, float]]],
+    ):
+        """Add predicted entity tags to CRF tokens."""
+        if ENTITY_ATTRIBUTE_TYPE in predictions:
+            _tags, _ = self._most_likely_tag(predictions[ENTITY_ATTRIBUTE_TYPE])
+            for tag, token in zip(_tags, crf_tokens):
+                token.entity_tag = tag
+
+    def _most_likely_tag(
+        self, predictions: List[Dict[Text, float]]
+    ) -> Tuple[List[Text], List[float]]:
+        """Get the entity tags with the highest confidence.
```

Comment:
in crf tags are found through the recurrent process that maximizes total sequence confidence. I thought, it is part of the crf

Reply:
The output of the CRF is a ranking of tags, so we get all possible tags with confidence values. This function just returns the tag with the highest confidence.